File: app/lib/data/scraper/parsers/category_normalizer.dart
import '../../../core/constants/app_constants.dart';

/// Maps Ottawa.ca raw activity labels to normalized swim types.
class SwimTypeNormalizer {
  static String normalize(String raw) {
    final lower = raw.toLowerCase().trim();

    if (_containsAny(lower, ['parent & tot', 'parent and tot', 'parent/tot'])) {
      return SwimCategories.parentTot;
    }
    if (_containsAny(lower, ['women', "women's", 'womens', 'female only'])) {
      return SwimCategories.womensSwim;
    }
    if (_containsAny(lower, ['preschool', 'pre-school', 'toddler'])) {
      return SwimCategories.preschoolSwim;
    }
    if (_containsAny(lower, ['teen swim', 'youth swim', 'teen only'])) {
      return SwimCategories.teenSwim;
    }
    if (_containsAny(lower, [
      'lane swim',
      'adult lane',
      'fitness swim',
      'length swim',
      'half lane',
    ])) {
      return SwimCategories.laneSwim;
    }
    if (_containsAny(lower, [
      'aquafit',
      'aqua fit',
      'aquafitness',
      'aqua-fit',
      'deep water aquafit',
      'deep-water aquafit',
    ])) {
      return SwimCategories.aquafit;
    }
    if (_containsAny(lower, [
      'aqua therapy',
      'aquatic therapy',
      'therapeutic',
      'therapy swim',
      'rehab swim',
      'rehabilitation swim',
      'sensory swim',
      'chronic pain',
      'pain management',
    ])) {
      return SwimCategories.therapeuticSwim;
    }
    if (_containsAny(lower, ['50+', '50 plus', 'fifty plus', 'senior swim'])) {
      return SwimCategories.adultSwim;
    }
    if (lower.contains('family')) {
      return SwimCategories.familySwim;
    }
    if (_containsAny(lower, ['adult swim', 'adults only swim', 'adult only'])) {
      return SwimCategories.adultSwim;
    }
    if (_containsAny(lower, [
      'public swim',
      'leisure swim',
      'recreation swim',
      'rec swim',
      'open swim',
      'general swim',
      'alternate needs swim',
      'accessibility swim',
      'accessible swim',
    ])) {
      return SwimCategories.generalSwim;
    }
    if (_containsAny(lower, ['wave swim', 'wave tank']) || lower == 'wave') {
      return SwimCategories.waveSwim;
    }
    if (lower.contains('swim')) {
      return SwimCategories.generalSwim;
    }

    return 'other';
  }

  static bool isSwimRow(String raw) {
    if (normalize(raw) != 'other') return true;
    final lower = raw.toLowerCase();
    return lower.contains('swim') ||
        lower.contains('aquafit') ||
        lower.contains('aqua fit') ||
        lower.contains('aqua therapy') ||
        lower.contains('aquatic') ||
        lower.contains('chronic pain') ||
        lower.contains('pain management') ||
        lower.contains('hot tub') ||
        lower.contains('wave') ||
        lower.contains('wading') ||
        lower.contains('splash');
  }

  static bool isDisplayable(String category) =>
      SwimCategories.all.contains(SwimCategories.normalizeStored(category));

  static bool _containsAny(String haystack, List<String> needles) =>
      needles.any(haystack.contains);

  /// Strips Ottawa.ca footnotes (e.g. "*Reservations required") from row labels.
  static String cleanRawActivityLabel(String raw) {
    var cleaned = raw.replaceAll(RegExp(r'[\u2018\u2019]'), "'");
    cleaned = cleaned.split(RegExp(r'[\r\n]+')).first;
    cleaned = cleaned.split('*').first;
    return cleaned.replaceAll(RegExp(r'\s+'), ' ').trim();
  }
}

String normalizeSwimCategory(String raw) => SwimTypeNormalizer.normalize(raw);

bool isSwimRow(String raw) => SwimTypeNormalizer.isSwimRow(raw);

File: scraper/ottawa_scraper/parsers/category_normalizer.py
"""Normalize swim category labels from Ottawa.ca."""

from __future__ import annotations

CANONICAL_CATEGORIES = {
    "general_swim",
    "lane_swim",
    "family_swim",
    "parent_tot",
    "aquafit",
    "adult_swim",
    "womens_swim",
    "therapeutic_swim",
    "wave_swim",
    "preschool_swim",
    "other",
}


def _contains_any(haystack: str, needles: tuple[str, ...]) -> bool:
    return any(n in haystack for n in needles)


def normalize_category(raw: str) -> str:
    lower = raw.lower().strip()

    if _contains_any(lower, ("parent & tot", "parent and tot", "parent/tot")):
        return "parent_tot"
    if _contains_any(lower, ("women", "women's", "womens", "female only")):
        return "womens_swim"
    if _contains_any(lower, ("preschool", "pre-school")):
        return "preschool_swim"
    if _contains_any(lower, ("lane swim", "adult lane", "fitness swim", "length swim")):
        return "lane_swim"
    if _contains_any(lower, ("aquafit", "aqua fit", "aquafitness", "aqua-fit")):
        return "aquafit"
    if _contains_any(lower, ("aqua therapy", "aquatic therapy", "therapeutic", "therapy swim")):
        return "therapeutic_swim"
    if "family" in lower:
        return "family_swim"
    if _contains_any(lower, ("adult swim", "adults only swim", "adult only")):
        return "adult_swim"
    if _contains_any(
        lower,
        (
            "public swim",
            "leisure swim",
            "recreation swim",
            "rec swim",
            "open swim",
            "general swim",
            "alternate needs swim",
            "accessibility swim",
            "accessible swim",
        ),
    ):
        return "general_swim"
    if _contains_any(lower, ("wave swim", "wave tank")) or lower == "wave":
        return "wave_swim"
    if "swim" in lower:
        return "general_swim"
    return "other"


def is_swim_row(raw: str) -> bool:
    if normalize_category(raw) != "other":
        return True
    lower = raw.lower()
    return any(
        kw in lower
        for kw in ("swim", "aquafit", "aqua fit", "aqua therapy", "hot tub", "wave")
    )
